### PR TITLE
Check unregistered needles when failing assert_screen

### PR DESCRIPTION
**Pull request by coolo**
https://progress.opensuse.org/issue/820

### CODE REVIEW

File: needle.pm [hunk 1]
```diff
@@ -203,13 +200,10 @@ sub has_property {
...
-        tags       => $self->{tags},
-        properties => $self->{properties},
-        area       => $self->{area},
-        file       => $self->{file},
-        png        => $self->{png},
-        name       => $self->{name}};
+    my $hash;
+    for my $key (qw(tags properties area file png unregistered name)) {
+        $hash->{$key} = $self->{$key};
+    }
```
Comment (okurz) [inline needle.pm]: @kraih would probably recommend

```perl
my $hash = map { $_ => $self->{$_} } qw(tags properties area file png unregistered name));
```
or something along these lines

File: needle.pm [hunk 2]
```diff
@@ -239,19 +233,20 @@ sub init {
...
-    if (!$goods || !@tags) {
+    if (!$goods || !@wanted) {
         return $goods || [];
     }
     my @results;
 
     # now check that it contains all the other tags too
   NEEDLE: for my $n (@$goods) {
-        for my $t (@tags) {
+        for my $t (@wanted) {
```
Comment (okurz) [inline needle.pm]: thank you for replacing the ambiguous `@tags` within the method `tags`. I got confused by it already :-)

File: backend/baseclass.pm [hunk 3]
```diff
@@ -856,6 +856,13 @@ sub check_asserted_screen {
             backend::baseclass::write_crash_file();
             bmwqemu::mydie "assert_screen fails, but we detected a timeout in the process, so we abort";
         }
+        elsif (!$self->assert_screen_check) {
+            my @unregistered_needles = grep { $_->{unregistered} } @{$self->assert_screen_needles};
+            my ($foundneedle, $candidates) = $img->search(\@unregistered_needles, 0, 1, undef);
+            # the best here is still a failure, as unregistered
+            push(@$failed_candidates, $foundneedle) if $foundneedle;
+            push(@$failed_candidates, @$candidates);
```
Comment (okurz) [inline backend/baseclass.pm]: well, that's basically what I wanted to say today in the call. Just add some `diag` and we already have something useful :-)

File: needle.pm (hunk 1 above)
Comment (kraih) [inline needle.pm]: Yes, i would. 😃 (`my %hash = map {...} qw(...)`)

Comment (coolo) [inline needle.pm]: actually it burned in my fingers :)

**Comment by coolo** [comment]
It's poo#820 - there is just a s missing in the expanded url

File: needle.pm (hunk 2 above)
Comment (AdamWill) [inline needle.pm]: Yes, now there's only the `tags()` method and the `%tags` hash ;)

**Comment by coolo** [comment]
this seems to miss a commit or two. At least unregistered is unused

**Comment by coolo** [comment]
but what is there is harmless so merging this

**Pull request merged**